variable/test/shape_test.cpp

```diff
@@ -155,3 +156,49 @@ TEST(VariableTest, reshape_mutable) {
   ASSERT_EQ(view, reference);
   ASSERT_EQ(var, modified_original);
 }
+
+TEST(ShapeTest, fold_x) {
+  const auto var = reshape(arange(Dim::X, 24), {{Dim::X, 6}, {Dim::Y, 4}});
+  const auto expected =
+      reshape(arange(Dim::X, 24), {{Dim::Row, 2}, {Dim::Time, 3}, {Dim::Y, 4}});
+  EXPECT_EQ(fold(var, Dim::X, {{Dim::Row, 2}, {Dim::Time, 3}}), expected);
+}
+
+TEST(ShapeTest, fold_y) {
+  const auto var = reshape(arange(Dim::X, 24), {{Dim::X, 6}, {Dim::Y, 4}});
+  const auto expected =
+      reshape(arange(Dim::X, 24), {{Dim::X, 6}, {Dim::Row, 2}, {Dim::Time, 2}});
+  EXPECT_EQ(fold(var, Dim::Y, {{Dim::Row, 2}, {Dim::Time, 2}}), expected);
+}
+
+TEST(ShapeTest, fold_into_3_dims) {
+  const auto var = arange(Dim::X, 24);
+  const auto expected =
+      reshape(arange(Dim::X, 24), {{Dim::Time, 2}, {Dim::Y, 3}, {Dim::Z, 4}});
+  EXPECT_EQ(fold(var, Dim::X, {{Dim::Time, 2}, {Dim::Y, 3}, {Dim::Z, 4}}),
+            expected);
+}
+
+TEST(ShapeTest, flatten) {
+  const auto var = reshape(arange(Dim::X, 24), {{Dim::X, 6}, {Dim::Y, 4}});
+  const auto expected = arange(Dim::Z, 24);
+  EXPECT_EQ(flatten(var, {Dim::X, Dim::Y}, Dim::Z), expected);
+}
+
+TEST(ShapeTest, flatten_only_2_dims) {
+  const auto var =
+      reshape(arange(Dim::X, 24), {{Dim::X, 2}, {Dim::Y, 3}, {Dim::Z, 4}});
+  const auto expected = reshape(arange(Dim::X, 24), {{Dim::X, 6}, {Dim::Z, 4}});
+  EXPECT_EQ(flatten(var, {Dim::X, Dim::Y}, Dim::X), expected);
+}
+
+TEST(ShapeTest, flatten_bad_dim_order) {
+  const auto var = reshape(arange(Dim::X, 24), {{Dim::X, 6}, {Dim::Y, 4}});
+  EXPECT_THROW(flatten(var, {Dim::Y, Dim::Z}, Dim::Z), except::DimensionError);
+}
```

Comment:
Hmmm, why is this bad dim order? There is no `Dim::Z` in the input. Did you mean `{Dim::Y, Dim::X}`?

Reply:
Good catch. Also highlights the bad design of the test: it still throws with a dimension error but not the right one...